libcache/cache.c

```diff
@@ -0,0 +1,650 @@
+};
+
+
+typedef struct cacheset_s {
+	cacheline_t *timestamps;
+	cacheline_t *tags[LIBCACHE_NUM_WAYS];
+	cacheline_t lines[LIBCACHE_NUM_WAYS];
+	size_t count;
+} cacheset_t;
+
+
+struct cachectx_s {
+	cacheset_t *sets;
+
+	size_t lineSize;
+	size_t numLines;
+	size_t numSets;
+
+	uint64_t tagMask;
+	uint64_t setMask;
+	uint64_t offMask;
+
+	uint8_t offBitsNum;
+	uint8_t setBitsNum;
+	uint8_t tagBitsNum;
+
+	cache_readCb_t readCb;
+	cache_writeCb_t writeCb;
+};
+
+
+/* Generates mask of type uint64_t with numBits set to 1 */
+static uint64_t cache_genMask(int numBits)
+{
+	return ((uint64_t)1 << numBits) - 1;
+}
+
+
+cachectx_t *cache_init(size_t size, size_t lineSize, cache_writeCb_t writeCb, cache_readCb_t readCb)
+{
+	cachectx_t *cache = NULL;
+
+	if (size == 0 || lineSize == 0 || size < lineSize) {
+		return NULL;
+	}
+
+	if (size % lineSize != 0 || (size / lineSize) % LIBCACHE_NUM_WAYS != 0) {
+		return NULL;
+	}
+
+	cache = malloc(sizeof(cachectx_t));
+
+	if (cache == NULL) {
+		return NULL;
+	}
+
+	cache->lineSize = lineSize;
+	cache->numLines = size / cache->lineSize;
+
+	cache->numSets = cache->numLines / LIBCACHE_NUM_WAYS;
+
+	cache->sets = calloc(cache->numSets, sizeof(cacheset_t));
+
+	if (cache->sets == NULL) {
+		free(cache);
+		return NULL;
+	}
+
+	cache->offBitsNum = LOG2(cache->lineSize);
+	cache->setBitsNum = LOG2(cache->numSets);
+	cache->tagBitsNum = LIBCACHE_ADDR_WIDTH - cache->setBitsNum - cache->offBitsNum;
+
+	cache->tagMask = cache_genMask(cache->tagBitsNum);
+	cache->setMask = cache_genMask(cache->setBitsNum);
+	cache->offMask = cache_genMask(cache->offBitsNum);
+
+	cache->readCb = readCb;
+	cache->writeCb = writeCb;
+
+	return cache;
+}
+
+
+int cache_deinit(cachectx_t *cache)
+{
+	int i, j;
+
+	for (i = 0; i < cache->numSets; ++i) {
+		for (j = 0; j < LIBCACHE_NUM_WAYS; ++j) {
+			if (IS_VALID(cache->sets[i].lines[j].flags)) {
+				free(cache->sets[i].lines[j].data);
+				CLEAR_VALID(cache->sets[i].lines[j].flags);
+			}
+		}
+	}
+	free(cache->sets);
+
+	free(cache);
+
+	return 0;
+}
+
+
+static int cache_compareTags(const void *lhs, const void *rhs)
+{
+	int ret = 0;
+
+	cacheline_t **lhsLine = (cacheline_t **)lhs;
+	cacheline_t **rhsLine = (cacheline_t **)rhs;
+
+	if (*lhsLine == NULL && *rhsLine == NULL) {
+		ret = 0;
+	}
+	else if (*lhsLine == NULL && *rhsLine != NULL) {
+		ret = -1;
+	}
+	else if (*lhsLine != NULL && *rhsLine == NULL) {
+		ret = 1;
+	}
+	else {
+		if ((*lhsLine)->tag < (*rhsLine)->tag) {
+			ret = -1;
+		}
+		else if ((*lhsLine)->tag == (*rhsLine)->tag) {
+			ret = 0;
+		}
+		else {
+			ret = 1;
+		}
+	}
+
+	return ret;
+}
+
+
+static ssize_t cache_flushLine(cache_writeCb_t writeCb, cacheline_t *linePtr, const uint64_t addr, size_t count)
+{
+	ssize_t written = 0;
+
+	if (linePtr != NULL && IS_VALID(linePtr->flags) && IS_DIRTY(linePtr->flags)) {
+		written = (*writeCb)(addr, linePtr->data, sizeof(*linePtr->data), count);
+		CLEAR_DIRTY(linePtr->flags);
+		if (written != count) {
```

Comment:
If written > 0 and != count we should do next write (destination write could not handle whole line at once). read/write wrapper to do this would be good